metagraph/src/common/vectors/vector_algorithm.hpp

```diff
@@ -78,42 +301,221 @@ inline bool fetch_and_unset_bit(uint64_t *v,
+        } else if (shift + width <= 16) {
+            // unaligned read from two bytes
+            return (__atomic_load_n((const uint16_t*)word, mo) >> shift) & mask;
+        } else if (shift + width <= 32) {
+            // unaligned read from four bytes
+            return (__atomic_load_n((const uint32_t*)word, mo) >> shift) & mask;
+        } else if (shift + width <= 64) {
+            // unaligned read from eight bytes
+            return (__atomic_load_n((const uint64_t*)word, mo) >> shift) & mask;
+        } else {
+            assert(false);
+        }
+    }
+
+    return 0;
+}
+
+inline uint64_t atomic_fetch_and_add(sdsl::int_vector<> &vector,
+                                     uint64_t i,
+                                     uint64_t val,
+                                     std::mutex &backup_mutex,
+                                     int mo) {
+    size_t width = vector.width();
+    size_t bit_pos = i * width;
+    uint64_t mask = (1llu << width) - 1;
+    if (width + 7 > 64) {
+        // there is no way to reliably modify without a mutex
+        std::lock_guard<std::mutex> lock(backup_mutex);
+        uint64_t old_val = vector[i];
+        vector[i] += val;
+        return old_val;
+    } else if ((bit_pos & 0x3F) + width <= 64) {
+        // element fits in an aligned word
+        uint64_t *word = &vector.data()[bit_pos >> 6];
+        uint8_t shift = bit_pos & 0x3F;
+        return (__atomic_fetch_add(word, val << shift, mo) >> shift) & mask;
+#if defined(MODE_TI) && defined(__CX16__)
+    } else if ((bit_pos & 0x7F) + width <= 128) {
+        // element fits in an aligned double word
+        __uint128_t *word = &reinterpret_cast<__uint128_t*>(vector.data())[bit_pos >> 7];
+        uint8_t shift = bit_pos & 0x7F;
+        // TODO: GCC only generates cmpxchg16b instruction only with older __sync functions
+        return (__sync_fetch_and_add(word, __uint128_t(val) << shift) >> shift) & mask;
+#endif
+    } else {
+        uint8_t shift = bit_pos & 0x7;
+        uint8_t *word = &reinterpret_cast<uint8_t*>(vector.data())[bit_pos >> 3];
+        if (shift + width <= 8) {
+            // read from a byte
+            return (__atomic_fetch_add(word, val << shift, mo) >> shift) & mask;
+        } else if (shift + width <= 16) {
+            // unaligned read from two bytes
+            return (__atomic_fetch_add((uint16_t*)word, val << shift, mo) >> shift) & mask;
+        } else if (shift + width <= 32) {
+            // unaligned read from four bytes
+            return (__atomic_fetch_add((uint32_t*)word, val << shift, mo) >> shift) & mask;
+        } else if (shift + width <= 64) {
+            // unaligned read from eight bytes
+            return (__atomic_fetch_add((uint64_t*)word, val << shift, mo) >> shift) & mask;
+        } else {
+            assert(false);
+        }
+    }
+
+    return 0;
+}
+
+inline uint64_t atomic_exchange(sdsl::int_vector<> &vector,
```

Comment:
In the GCC builtins, setting a new value and returning the old one is called `exchange` instead of `set`

Reply:
The closely related methods implemented for bit vectors are named fetch_and_(un)set because there's no variable passed with which the stored value should be swapped.

Reply:
I like `atomic_fetch_and_set`, but `exchange` is also fine.
I agree that `atomic_set` doesn't make it clear what is returned.